File: back-node/src/services/postValidator.ts
import {
  calculateReadingTime,
  validateContentBlocks,
  isValidCategory,
} from "../services/contentService.js";
import { CreatePostRequest, UpdatePostRequest } from "../types/content.js";

/**
 * Validar dados de criação de post
 * Corresponde à validação do formulário do frontend
 */
export function validateCreatePostData(data: any): {
  valid: boolean;
  errors: string[];
} {
  const errors: string[] = [];

  // Validar title
  if (!data.title || typeof data.title !== "string") {
    errors.push("title é obrigatório e deve ser string");
  } else if (data.title.trim().length === 0) {
    errors.push("title não pode estar vazio");
  } else if (data.title.length > 200) {
    errors.push("title não pode ter mais de 200 caracteres");
  }

  // Validar category
  if (!data.category || typeof data.category !== "string") {
    errors.push("category é obrigatória e deve ser string");
  } else if (!isValidCategory(data.category)) {
    errors.push(`category deve ser uma das opções válidas`);
  }

  // Validar tags
  if (data.tags) {
    if (!Array.isArray(data.tags)) {
      errors.push("tags deve ser um array");
    } else if (data.tags.length > 20) {
      errors.push("tags não pode ter mais de 20 itens");
    } else if (!data.tags.every((t: any) => typeof t === "string")) {
      errors.push("todas as tags devem ser strings");
    }
  }

  // Validar image
  if (
    !data.image ||
    (typeof data.image !== "string" && !isImageFile(data.image))
  ) {
    errors.push("image é obrigatória (URL ou arquivo)");
  } else if (typeof data.image === "string" && data.image.length > 500) {
    errors.push("URL da image não pode ter mais de 500 caracteres");
  }

  // Validar excerpt
  if (!data.excerpt || typeof data.excerpt !== "string") {
    errors.push("excerpt é obrigatória e deve ser string");
  } else if (data.excerpt.trim().length === 0) {
    errors.push("excerpt não pode estar vazia");
  } else if (data.excerpt.length > 500) {
    errors.push("excerpt não pode ter mais de 500 caracteres");
  }

  // Validar conteudo (validação desligada - apenas comentada)
  // if (!data.conteudo) {
  //   errors.push("conteudo é obrigatório");
  // } else if (!Array.isArray(data.conteudo)) {
  //   errors.push("conteudo deve ser um array");
  // } else if (data.conteudo.length === 0) {
  //   errors.push("conteudo deve ter pelo menos um bloco");
  // } else if (!validateContentBlocks(data.conteudo)) {
  //   errors.push("alguns blocos de conteudo são inválidos");
  // }

  // Validar readingTime (opcional)
  if (data.readingTime && typeof data.readingTime !== "string") {
    errors.push("readingTime deve ser string");
  }

  return {
    valid: errors.length === 0,
    errors,
  };
}

/**
 * Sanitizar dados de criação de post
 */
export function sanitizeCreatePostData(data: any): CreatePostRequest {
  return {
    title: String(data.title).trim().slice(0, 200),
    category: String(data.category),
    tags: Array.isArray(data.tags)
      ? data.tags.map((t: any) => String(t).trim()).slice(0, 20)
      : [],
    image: String(data.image).trim().slice(0, 500),
    excerpt: String(data.excerpt).trim().slice(0, 500),
    conteudo: Array.isArray(data.conteudo) ? data.conteudo : [],
    readingTime: data.readingTime ? String(data.readingTime).trim() : undefined,
  };
}

/**
 * Preparar dados para salvar no banco
 */
export function preparePostData(request: CreatePostRequest) {
  const readingTime =
    request.readingTime || calculateReadingTime(request.conteudo);

  return {
    title: request.title,
    category: request.category,
    tags: request.tags,
    image: request.image,
    excerpt: request.excerpt,
    conteudo: request.conteudo,
    reading_time: readingTime,
    date: new Date().toISOString(),
  };
}

/**
 * Validar dados de atualização de post
 */
export function validateUpdatePostData(data: any): {
  valid: boolean;
  errors: string[];
} {
  const errors: string[] = [];

  // Todos os campos são opcionais em UPDATE
  if (data.title !== undefined) {
    if (typeof data.title !== "string") {
      errors.push("title deve ser string");
    } else if (data.title.length > 200) {
      errors.push("title não pode ter mais de 200 caracteres");
    }
  }

  if (data.category !== undefined) {
    if (typeof data.category !== "string") {
      errors.push("category deve ser string");
    } else if (!isValidCategory(data.category)) {
      errors.push("category deve ser uma das opções válidas");
    }
  }

  if (data.tags !== undefined) {
    if (!Array.isArray(data.tags)) {
      errors.push("tags deve ser um array");
    } else if (data.tags.length > 20) {
      errors.push("tags não pode ter mais de 20 itens");
    }
  }

  // Validação de conteudo desligada (apenas comentada)
  // if (data.conteudo !== undefined) {
  //   if (!Array.isArray(data.conteudo)) {
  //     errors.push("conteudo deve ser um array");
  //   } else if (data.conteudo.length === 0) {
  //     errors.push("conteudo deve ter pelo menos um bloco");
  //   } else if (!validateContentBlocks(data.conteudo)) {
  //     errors.push("alguns blocos de conteudo são inválidos");
  //   }
  // }

  return {
    valid: errors.length === 0,
    errors,
  };
}

/**
 * Sanitizar dados de atualização
 */
export function sanitizeUpdatePostData(data: any): UpdatePostRequest {
  const sanitized: UpdatePostRequest = {};

  if (data.title !== undefined) {
    sanitized.title = String(data.title).trim().slice(0, 200);
  }
  if (data.category !== undefined) {
    sanitized.category = String(data.category);
  }
  if (data.tags !== undefined) {
    sanitized.tags = Array.isArray(data.tags)
      ? data.tags.map((t: any) => String(t).trim()).slice(0, 20)
      : [];
  }
  if (data.image !== undefined) {
    sanitized.image = String(data.image).trim().slice(0, 500);
  }
  if (data.excerpt !== undefined) {
    sanitized.excerpt = String(data.excerpt).trim().slice(0, 500);
  }
  if (data.conteudo !== undefined) {
    sanitized.conteudo = Array.isArray(data.conteudo) ? data.conteudo : [];
  }
  if (data.readingTime !== undefined) {
    sanitized.readingTime = String(data.readingTime).trim();
  }

  return sanitized;
}

// ===========================
// Utilitários privados
// ===========================

function isImageFile(file: any): boolean {
  return file && typeof file === "object" && file.path && file.filename;
}

File: back-java/src/main/java/com/blog/service/PostValidator.java
package com.blog.service;

import com.blog.model.CreatePostRequest;
import com.blog.model.UpdatePostRequest;
import org.springframework.stereotype.Service;


import java.util.Map;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class PostValidator {

    private final ContentService contentService;

    public PostValidator(ContentService contentService) {
        this.contentService = contentService;
    }

    public ValidationResult validateCreatePostData(Object title, Object category, Object tags,
                                                   Object image, Object excerpt, Object conteudo, Object readingTime) {
        List<String> errors = new ArrayList<>();

        if (title == null || !(title instanceof String) || ((String) title).trim().isEmpty()) {
            errors.add("title é obrigatório e deve ser string");
        } else if (((String) title).length() > 200) {
            errors.add("title não pode ter mais de 200 caracteres");
        }

        if (category == null || !(category instanceof String)) {
            errors.add("category é obrigatória e deve ser string");
        } else if (!contentService.isValidCategory((String) category)) {
            errors.add("category deve ser uma das opções válidas");
        }

        if (tags != null) {
            if (!(tags instanceof List)) {
                errors.add("tags deve ser um array");
            } else if (((List<?>) tags).size() > 20) {
                errors.add("tags não pode ter mais de 20 itens");
            } else if (((List<?>) tags).stream().anyMatch(t -> !(t instanceof String))) {
                errors.add("todas as tags devem ser strings");
            }
        }

        if (image == null || (image instanceof String && ((String) image).isBlank())) {
            errors.add("image é obrigatória (URL ou arquivo)");
        } else if (image instanceof String && ((String) image).length() > 500) {
            errors.add("URL da image não pode ter mais de 500 caracteres");
        }

        if (excerpt == null || !(excerpt instanceof String) || ((String) excerpt).trim().isEmpty()) {
            errors.add("excerpt é obrigatória e deve ser string");
        } else if (((String) excerpt).length() > 500) {
            errors.add("excerpt não pode ter mais de 500 caracteres");
        }

        if (conteudo == null) {
            errors.add("conteudo é obrigatório");
        } else if (!(conteudo instanceof List)) {
            errors.add("conteudo deve ser um array");
        } else {
            @SuppressWarnings("unchecked")
            List<Object> list = (List<Object>) conteudo;
            if (list.isEmpty()) {
                errors.add("conteudo deve ter pelo menos um bloco");
            } else {
                List<Map<String, Object>> blocks = list.stream()
                        .filter(m -> m instanceof Map)
                        .map(m -> (Map<String, Object>) m)
                        .collect(Collectors.toList());
                // Validação de blocos de conteúdo desligada – aceita qualquer estrutura
                // if (blocks.size() != list.size() || !contentService.validateContentBlocks(blocks)) {
                //     errors.add("alguns blocos de conteudo são inválidos");
                // }
            }
        }

        if (readingTime != null && !(readingTime instanceof String)) {
            errors.add("readingTime deve ser string");
        }

        return new ValidationResult(errors.isEmpty(), errors);
    }

    public ValidationResult validateUpdatePostData(Object title, Object category, Object tags,
                                                   Object image, Object excerpt, Object conteudo, Object readingTime) {
        List<String> errors = new ArrayList<>();

        if (title != null) {
            if (!(title instanceof String)) {
                errors.add("title deve ser string");
            } else if (((String) title).length() > 200) {
                errors.add("title não pode ter mais de 200 caracteres");
            }
        }

        if (category != null && !contentService.isValidCategory(category.toString())) {
            errors.add("category deve ser uma das opções válidas");
        }

        if (tags != null && !(tags instanceof List)) {
            errors.add("tags deve ser um array");
        } else if (tags instanceof List && ((List<?>) tags).size() > 20) {
            errors.add("tags não pode ter mais de 20 itens");
        }

        if (conteudo != null) {
            if (!(conteudo instanceof List)) {
                errors.add("conteudo deve ser um array");
            } else {
                List<?> list = (List<?>) conteudo;
                if (list.isEmpty()) {
                    errors.add("conteudo deve ter pelo menos um bloco");
                } else {
                    List<Map<String, Object>> blocks = list.stream()
                            .filter(m -> m instanceof Map)
                            .map(m -> (Map<String, Object>) m)
                            .collect(Collectors.toList());
                    // Validação de blocos de conteúdo desligada – aceita qualquer estrutura
                    // if (blocks.size() != list.size() || !contentService.validateContentBlocks(blocks)) {
                    //     errors.add("alguns blocos de conteudo são inválidos");
                    // }
                }
            }
        }

        return new ValidationResult(errors.isEmpty(), errors);
    }

    public CreatePostRequest sanitizeCreatePostData(String title, String category, List<String> tags,
                                                    String image, String excerpt, List<Map<String, Object>> conteudo,
                                                    String readingTime) {
        CreatePostRequest r = new CreatePostRequest();
        r.setTitle(title != null ? title.trim().substring(0, Math.min(200, title.trim().length())) : "");
        r.setCategory(category != null ? category : "");
        r.setTags(tags != null ? tags.stream().map(t -> t != null ? t.trim() : "").limit(20).collect(Collectors.toList()) : List.of());
        r.setImage(image != null ? image.trim().substring(0, Math.min(500, image.trim().length())) : "");
        r.setExcerpt(excerpt != null ? excerpt.trim().substring(0, Math.min(500, excerpt.trim().length())) : "");
        r.setConteudo(conteudo != null ? conteudo : List.of());
        r.setReadingTime(readingTime != null ? readingTime.trim() : null);
        return r;
    }

    public UpdatePostRequest sanitizeUpdatePostData(String title, String category, List<String> tags,
                                                    String image, String excerpt, List<Map<String, Object>> conteudo,
                                                    String readingTime) {
        UpdatePostRequest r = new UpdatePostRequest();
        if (title != null) r.setTitle(title.trim().substring(0, Math.min(200, title.trim().length())));
        if (category != null) r.setCategory(category);
        if (tags != null) r.setTags(tags.stream().map(t -> t != null ? t.trim() : "").limit(20).collect(Collectors.toList()));
        if (image != null) r.setImage(image.trim().substring(0, Math.min(500, image.trim().length())));
        if (excerpt != null) r.setExcerpt(excerpt.trim().substring(0, Math.min(500, excerpt.trim().length())));
        if (conteudo != null) r.setConteudo(conteudo);
        if (readingTime != null) r.setReadingTime(readingTime.trim());
        return r;
    }

    public record ValidationResult(boolean valid, List<String> errors) {}
}
